src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -238,19 +290,55 @@ class Wavefunction
         rng_.seed(s);
     }
 
+    void reorderWFN(unsigned qubitLoc, unsigned newPos) const
+    {
+        //get id of qubit located at newPos and qubitLoc - getting index from the element
+        auto itr1 = std::find(qubitmap_.begin(), qubitmap_.end(), newPos);
+        auto newPosId = std::distance(qubitmap_.begin(), itr1);
+        auto itr2 = std::find(qubitmap_.begin(), qubitmap_.end(), qubitLoc);
+        auto qid = std::distance(qubitmap_.begin(), itr2);
+        //swap qubits in wfn between qubitLoc and newPos
+        if (newPos != qubitLoc)
+        {
+            for (unsigned i = 0; i < wfn_.size(); i++)
+            {
+                unsigned int bit1 = (i >> qubitLoc) & 1;
+                unsigned int bit2 = (i >> newPos) & 1;
+                unsigned int x = (bit1 ^ bit2);
+                x = (x << qubitLoc) | (x << newPos);
+                unsigned new_i = i ^ x;
+                if (new_i > i)
+                {
+                    std::iter_swap(wfn_.begin() + i, wfn_.begin() + new_i);
+                }
```

Comment:
This is a complicated piece of math that is easy to get wrong. Have you considered adding unit tests that check this logic directly?

Reply:
The tests currently present test the reordering routine, thereby confirming the correctness of this bit math. However, as far as testing this directly I am unsure of how that could be done, as this method is deeply embeded making it very difficult to access outside the class

Reply:
A common way to improve testability is to extract the functionality to be tested into a standalone function or a helper class with public methods to expose the code. The function/class then can be tested with given parameters in known state, while the production class will use them with its internal data. This way you can ensure much better test coverage and easier debugging of failures.

Reply:
Given the lack of any testing infrastructure in the native simulator, I want to punt this to after I have a discussion with Dave and Thomas. But will definitely keep this in mind!

Reply:
Why not add a test to local_test.cpp? I agree that the test infrastructure in the native simulator is far from being convenient, but it does allow to create instances of Wavefunction class and execute public methods on them, which would give you direct access to the code you are adding. Given the complexity of the math here I am not comfortable proving its correctness by review and system-level tests only.